File: Ref_Code/基于STM32的锁相放大器/基于STM32的锁相放大器/IQ检波/Lockin_Amplifier.c
#include "Lockin_Amplifier.h"
#include "arm_math.h"
__attribute__((section("ccmram"))) static float SinLUT[LUT_SIZE];
#define INV_256 0.00390625f

void DLIA_Init_LUT(void)
{
    for(int i = 0; i < LUT_SIZE; i++)
    {
        float angle = (2.0f * PI * (float)i) / (float)LUT_SIZE;
        SinLUT[i] = arm_sin_f32(angle);
    }
}

void DLIA_Init(DLIA_Config_t* config, DLIA_State_t* state,float freq , float fs , float lpf_beta)
{
    config->Signal_Freq = freq;
    config->Ref_Freq = fs;
    config->LPF_Beta = lpf_beta;
	
    state->PhaseStep = (uint32_t)(config->Signal_Freq / config->Ref_Freq * 4294967296.0f); 
    state->PhaseAccumulator = 0;

    state->LPF_Val_I = 0.0f;
    state->LPF_Val_Q = 0.0f;

}

void DLIA_Process(DLIA_Config_t* config, DLIA_State_t* state, DLIA_Output_t* output, uint16_t* ADCValue, uint32_t length)
{
    if(length != DLIA_BLOCK_SIZE)
    {
        return;
    }

    uint32_t phase_acc  = state->PhaseAccumulator;
    uint32_t phase_step = state->PhaseStep;

    float sum_I = 0.0f;
    float sum_Q = 0.0f;

    const float k_convert =  3.3f * 0.000244140625f;  // ADC值转电压: 3.3V参考/4096
    const uint32_t cos_offset = LUT_SIZE >> 2;  // 余弦表偏移: LUT_SIZE/4 (90度相位差)

    float sum_ref_sin = 0.0f;  // 参考正弦累加和
    float sum_ref_cos = 0.0f;  // 参考余弦累加和
    uint32_t sum_adc_raw = 0;  // 原始ADC值累加和 

    for(int i = 0; i < length; i++)
    {
        uint32_t phase_fixed = phase_acc >> (LUT_SHIFT - 8); 

        uint32_t idx0 = (phase_fixed >> 8) & LUT_MASK;  
        uint32_t idx1 = (idx0 + 1) & LUT_MASK;  

        float frac = (float)(phase_fixed & 0xFF) * INV_256;  // 插值分数 (0-1)
        float frac_inv = 1.0f - frac;  // 1-分数

        float val_sin0 = SinLUT[idx0];
        float val_sin1 = SinLUT[idx1];
        float ref_sin  = val_sin0 * frac_inv + val_sin1 * frac;  // 正弦参考信号线性插值

        uint32_t idx_cos0 = (idx0 + cos_offset) & LUT_MASK;  // 余弦索引0
        uint32_t idx_cos1 = (idx1 + cos_offset) & LUT_MASK;  // 余弦索引1

        float val_cos0 = SinLUT[idx_cos0];
        float val_cos1 = SinLUT[idx_cos1];
        float ref_cos  = val_cos0 * frac_inv + val_cos1 * frac;  // 余弦参考信号线性插值

        phase_acc += phase_step;

        float adc_val = (float)ADCValue[i];
        sum_I += adc_val * ref_sin;  // I通道解调
        sum_Q += adc_val * ref_cos;  // Q通道解调

        sum_adc_raw += ADCValue[i];  // 记录原始ADC值用于直流消除
        sum_ref_sin += ref_sin;      // 记录参考信号用于归一化
        sum_ref_cos += ref_cos;     

    }
    state->PhaseAccumulator = phase_acc;

    float avg_adc_raw = (float)sum_adc_raw / (float)length;  // 计算直流分量

    // 消除直流偏移
    float final_sum_I = sum_I - (avg_adc_raw * sum_ref_sin);
    float final_sum_Q = sum_Q - (avg_adc_raw * sum_ref_cos);

    // 归一化处理
    float norm_factor = k_convert * (2.0f / (float)length);
    float avg_I = final_sum_I * norm_factor;
    float avg_Q = final_sum_Q * norm_factor;

    // 一阶低通滤波器
    state->LPF_Val_I += config->LPF_Beta * (avg_I - state->LPF_Val_I);
    state->LPF_Val_Q += config->LPF_Beta * (avg_Q - state->LPF_Val_Q);

    output->Output_I = state->LPF_Val_I;
    output->Output_Q = state->LPF_Val_Q;

    // 计算幅度: sqrt(I^2 + Q^2)
    float sum = output->Output_I * output->Output_I + output->Output_Q * output->Output_Q;
    arm_sqrt_f32(sum, &output->Output_Amp);

    // 计算相位: atan2(Q, I)
    float phase;
    phase = atan2f(output->Output_Q, output->Output_I);

    output->Output_PhaseRad = phase;
    output->Output_PhaseDeg = phase * 57.2957795f;  // 弧度转角度

}

File: Ref_Code/基于STM32的锁相放大器/基于STM32的锁相放大器/MATLAB仿真/Lockin_Amplifier.m
classdef Lockin_Amplifier
    properties (Constant)
        PI = 3.14159265358979;
        DLIA_BLOCK_SIZE = 512;         % 每次处理的样本数
        LUT_SIZE = 1024;               % LUT表大小
        LUT_MASK = 1023;               % LUT掩码 (LUT_SIZE - 1)
        LUT_SHIFT = 22;                % LUT索引右移位数 (32 - log2(LUT_SIZE))
    end
    
    properties (Access = private)
        SinLUT;                        % 正弦查找表
    end
    
    methods
        function obj = Lockin_Amplifier()
            % 构造函数，初始化LUT
            obj = obj.DLIA_Init_LUT();
        end
        
        function obj = DLIA_Init_LUT(obj)
            % 初始化正弦查找表
            obj.SinLUT = zeros(1, obj.LUT_SIZE);
            
            for i = 0:obj.LUT_SIZE-1
                angle = (2.0 * obj.PI * i) / obj.LUT_SIZE;
                obj.SinLUT(i+1) = sin(angle);  % MATLAB索引从1开始
            end
        end
        
        function [config, state] = DLIA_Init(obj, freq, fs, lpf_beta, dc_offset)
            % 初始化配置和状态
            config.Signal_Freq = freq;
            config.Ref_Freq = fs;
            config.LPF_Beta = lpf_beta;
            config.DC_Offset = dc_offset; % 这个值现在仅作为初始观察，算法内部会自动计算真实DC
            
            state.PhaseStep = uint32(freq / fs * 4294967296.0);
            state.PhaseAccumulator = uint32(0);
            state.LPF_Val_I = 0.0;
            state.LPF_Val_Q = 0.0;
        end
        
        function [state, output] = DLIA_Process(obj, config, state, ADCValue)
            % 处理ADC数据
            len = numel(ADCValue);
            
            if len ~= obj.DLIA_BLOCK_SIZE
                error('输入数据长度必须等于DLIA_BLOCK_SIZE');
            end
            
            % 加载状态
            phase_acc = state.PhaseAccumulator;
            phase_step = state.PhaseStep;
            
            % 用于 "Sum(Raw * Ref)"
            sum_I_raw = 0.0;
            sum_Q_raw = 0.0;
            
            % 用于 "协方差修正" 的辅助统计变量
            sum_adc_raw = 0.0; % Sum(Raw)
            sum_sin_ref = 0.0; % Sum(Sin)
            sum_cos_ref = 0.0; % Sum(Cos)
            
            k_convert = 3.3 * 0.000244140625;  % 3.3V / 4096
            
            % 预计算常量
            cos_offset = bitshift(obj.LUT_SIZE, -2);  % 256
            INV_256 = 1.0 / 256.0;
            
            for i = 1:len
                phase_fixed = bitshift(phase_acc, -(obj.LUT_SHIFT - 8)); % 保留8位小数
                
                % 获取0基准的索引
                idx0_base = bitand(bitshift(phase_fixed, -8), obj.LUT_MASK); 
                idx1_base = bitand(idx0_base + 1, obj.LUT_MASK);
                
                % MATLAB索引+1
                idx0 = idx0_base + 1;
                idx1 = idx1_base + 1;
                
                frac = double(bitand(phase_fixed, 255)) * INV_256;
                frac_inv = 1.0 - frac;
                
                % Sin 插值
                val_sin0 = obj.SinLUT(idx0);
                val_sin1 = obj.SinLUT(idx1);
                ref_sin = val_sin0 * frac_inv + val_sin1 * frac;
                
                % Cos 插值
                idx_cos0 = bitand(idx0_base + cos_offset, obj.LUT_MASK) + 1;
                idx_cos1 = bitand(idx1_base + cos_offset, obj.LUT_MASK) + 1;
                
                val_cos0 = obj.SinLUT(idx_cos0);
                val_cos1 = obj.SinLUT(idx_cos1);
                ref_cos = val_cos0 * frac_inv + val_cos1 * frac;
                
                phase_acc = phase_acc + phase_step; % Matlab uint32会自动处理溢出吗？
                 phase_acc_dbl = double(phase_acc) + double(phase_step);
                 phase_acc = uint32(mod(phase_acc_dbl, 4294967296));
                
                val_adc_float = double(ADCValue(i)); % 0~4095
                
                sum_I_raw = sum_I_raw + (val_adc_float * ref_sin);
                sum_Q_raw = sum_Q_raw + (val_adc_float * ref_cos);
                
                sum_adc_raw = sum_adc_raw + val_adc_float;
                sum_sin_ref = sum_sin_ref + ref_sin;
                sum_cos_ref = sum_cos_ref + ref_cos;
            end
            
            state.PhaseAccumulator = phase_acc;

            avg_adc_raw_dc = sum_adc_raw / len;
            
            final_sum_I = sum_I_raw - (avg_adc_raw_dc * sum_sin_ref);
            final_sum_Q = sum_Q_raw - (avg_adc_raw_dc * sum_cos_ref);
            
            scale_factor = (2.0 / obj.DLIA_BLOCK_SIZE) * k_convert;
            
            avg_I = final_sum_I * scale_factor;
            avg_Q = final_sum_Q * scale_factor;
            
            state.LPF_Val_I = state.LPF_Val_I + config.LPF_Beta * (avg_I - state.LPF_Val_I);
            state.LPF_Val_Q = state.LPF_Val_Q + config.LPF_Beta * (avg_Q - state.LPF_Val_Q);
            
            output.Output_I = state.LPF_Val_I;
            output.Output_Q = state.LPF_Val_Q;
            
            sum_sq = output.Output_I^2 + output.Output_Q^2;
            output.Output_Amp = sqrt(sum_sq);
            

            phase = atan2(output.Output_Q, output.Output_I);
            output.Output_PhaseRad = phase;
            output.Output_PhaseDeg = phase * 57.2957795;
        end
    end
end